File: src/test/java/com/laamella/code_state_machine/PriorityTests.java
package com.laamella.code_state_machine;

import static com.laamella.code_state_machine.builder.DslStateMachineBuilder.always;
import static com.laamella.code_state_machine.builder.DslStateMachineBuilder.never;
import static com.laamella.code_state_machine.util.SimpleState.A;
import static com.laamella.code_state_machine.util.SimpleState.B;
import static org.junit.Assert.assertEquals;

import org.junit.Before;
import org.junit.Test;

import com.laamella.code_state_machine.builder.DslStateMachineBuilder;
import com.laamella.code_state_machine.priority.Priority;
import com.laamella.code_state_machine.util.SimpleState;

public class PriorityTests {
	private final StringBuffer trace = new StringBuffer();

	private DslStateMachineBuilder<SimpleState, Object, Priority> machineBuilder;

	@Before
	public void before() {
		machineBuilder = new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			{
				state(A).isAStartState();
				state(B).isAnEndState();
			}
		};
	}

	private TraceAction trace(final String signature) {
		return new TraceAction(trace, signature);
	}

	@Test
	public void highPrioIsTheOnlyOneFiring() {
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.LOWEST, trace("L"));

		machineBuilder.buildMachine().poll();
		assertEquals("H", trace.toString());
	}

	@Test
	public void normalPriosAreTheOnlyOnesFiringBecauseOtherPrioDoesntMeetCondition() {
		machineBuilder.state(A).when(always()).transition(B, never(), Priority.HIGH, trace("H"));
		machineBuilder.state(A).when(always()).transition(B, never(), Priority.NORMAL, trace("N"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));

		machineBuilder.buildMachine().poll();
		assertEquals("NN", trace.toString());
	}

	@Test
	public void equalPriosFireTogether() {
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
		machineBuilder.state(A).when(always()).transition(B, always(), Priority.LOWEST, trace("L"));

		machineBuilder.buildMachine().poll();
		assertEquals("HH", trace.toString());
	}

}

File: scala/src/test/scala/com/laamella/code_state_machine/PriorityTests.scala
package com.laamella.code_state_machine

import com.laamella.code_state_machine.builder.DslStateMachineBuilder
import com.laamella.code_state_machine.priority.LeveledPriority._
import com.laamella.code_state_machine.util.{A, B, SimpleState}
import org.scalatest.BeforeAndAfterEach

class PriorityTests extends UnitSpec with BeforeAndAfterEach {
  private var trace: StringBuilder = _

  var builder: DslStateMachineBuilder[SimpleState, Object, Value] = _

  override def beforeEach() {
    builder = new DslStateMachineBuilder[SimpleState, Object, Value](NORMAL)
    builder.state(A).isAStartState
    builder.state(B).isAnEndState

    trace = new StringBuilder()
  }

  private def trace(signature: String): TraceAction = {
    new TraceAction(trace, signature)
  }

  behavior of "transitions with priorities"

  they should "only fire when they have the highest priority" in {
    builder.state(A).always.withPrio(HIGH).doing(trace("H")).goTo(B)
    builder.state(A).always.withPrio(NORMAL).doing(trace("N")).goTo(B)
    builder.state(A).always.withPrio(LOWEST).doing(trace("L")).goTo(B)
    val machine = builder.build()

    machine.poll()

    assert("H" == trace.toString)
  }

  they should "only fire when their condition is met, even though higher priority transactions exist" in {
    builder.state(A).never.withPrio(HIGH).doing(trace("H")).goTo(B)
    builder.state(A).never.withPrio(NORMAL).doing(trace("N")).goTo(B)
    builder.state(A).always.withPrio(NORMAL).doing(trace("N")).goTo(B)
    builder.state(A).always.withPrio(NORMAL).doing(trace("N")).goTo(B)
    val machine = builder.build()

    machine.poll()

    assert("NN" == trace.toString)
  }

  they should "fire together when they have the same priority" in {
    builder.state(A).always.doing(trace("H")).withPrio(HIGH).goTo(B)
    builder.state(A).always.doing(trace("H")).withPrio(HIGH).goTo(B)
    builder.state(A).always.doing(trace("L")).withPrio(LOWEST).goTo(B)
    val machine = builder.build()

    machine.poll()
    assert("HH" == trace.toString)
  }

}
